### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by jquense**

### CODE REVIEW

**Review by itajaja — COMMENTED** [review]
uuuuh what does this do?

File: logify.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+const isPlainObject = require('lodash/isPlainObject');
+const { logger: _logger } = require('./index');
+
+function formatValue(arg) {
+  if (typeof arg === 'function') {
+    return `ƒ() ${arg.name} {...}`;
```
Comment (taion) [inline logify.js]: fancy

File: package.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  "version": "1.0.0",
+  "main": "index.js",
+  "repository": "https://github.com/4Catalyzer/logging.git",
+  "author": "4Catalyzer",
+  "license": "MIT",
+  "scripts": {
+    "lint":
+      "eslint . && prettier --list-different --ignore-path .ignore '**/*.{json,css,md}'",
+    "format":
+      "eslint . --fix && prettier --write --ignore-path .eslintignore '**/*.{json,css,md}'",
```
Comment (taion) [inline package.json:12]: shouldn't it be `eslint --fix .`? or is this just the scaffolding? :p

File: logify.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+function formatArguments(args) {
+  const fullArgs = args.map(formatValue).join(', ');
+  return fullArgs.length > 100 ? `${fullArgs.slice(0, 100)}...` : fullArgs;
+}
+
+function formatMethod(method, args) {
+  return `${method}(${formatArguments(args)})`;
+}
+
+module.exports = function logify(target, property, desc, level = 'verbose') {
```
Comment (taion) [inline logify.js]: so you can use either `@logify` or `@logify('info')`?

Comment (jquense) [reply logify.js]: ya

**Comment by jquense** [comment]
sorry, mostly preconfigures winston a bit, and makes the API less annoying to work with for adding groups.

Also adds a decorator for instrumenting methods, we use something like it in factory software app

**Comment by jquense** [comment]
cc @yodaiken as well. I just sort of assuming it's worth having a common config here. we can keep adding/tweaking

File: index.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+const omit = require('lodash/omit');
```
Comment (itajaja) [inline index.js]: :( https://github.com/lodash/lodash/issues/2930

Comment (jquense) [reply index.js]: WTF, that's so unlike lodash to be so opinionated about that. I don't think when you're the defacto standard library of a language you should be able to tell everyone with there infinite use-cases that common pattern foo is not good.

Comment (itajaja) [reply index.js]: but computers...

File: index.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  const meta = omit(i, ['level', 'message', 'label']);
+  if (!Object.keys(meta).length) return i;
+  i.message = `${i.message}\t${JSON.stringify(meta, null, 2)}`; // eslint-disable-line
+  return i;
+});
+
+class ConsoleTransport extends transports.Console {
+  constructor(opts) {
+    super(opts);
+    this.silent = opts.silent;
```
Comment (taion) [inline index.js:29]: should `silent` be spread out so it doesn't go to the super ctor?

File: index.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+const { combine, colorize, label: addLabel, printf } = format;
+
+const container = new Container();
+const TEST = process.env.NODE_ENV === 'test';
+
+let LEVEL = 'info';
+
+const appendMeta = format(i => {
+  const meta = omit(i, ['level', 'message', 'label']);
```
Comment (taion) [inline index.js]: yeah but you can just write this as:

```js
const { level: _1, message: _2, label: _3, ...meta } = i;
```

it's not _that_ bad, and it's actually a ton faster than `omit()` too

File: index.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+function add(id, label = id) {
+  container.add(id, {
+    level: LEVEL,
+    format: combine(
+      ...[
+        colorize(),
+        label && addLabel({ label, message: true }),
+        appendMeta(),
+        printf(i => `${i.level}:  ${i.message}`),
```
Comment (taion) [inline index.js]: two spaces between level and message?

File: index.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  log(...args) {
+    if (this.silent) return;
+    super.log(...args);
+  }
+}
+
+function add(id, label = id) {
+  container.add(id, {
+    level: LEVEL,
```
Comment (taion) [inline index.js]: wdyt of doing

```js
get level: () => LEVEL,
```

instead of having to loop through all containers below when changing level?

**Review by taion — APPROVED** [review]
see minor notes

File: index.js (hunk 5 above)
Comment (jquense) [reply index.js:29]: its _supposed_ to be supported here so i'm hoping well just remove this at some point

File: index.js (hunk 8 above)
Comment (jquense) [reply index.js]: i can try, not sure if it'll work, any Object.assign along the way and it'll kill the getters

Comment (taion) [reply index.js]: ok fair enough just trying to be cute, no big deal if it's not robust

File: logify.js (hunk 1 above)
Comment (yodaiken) [reply logify.js]: is it kosher for us to use unicode in source code now ?

Comment (jquense) [reply logify.js]: i mean it's not python ;)

Comment (jquense) [reply logify.js]: seriously tho, i do it in JS b/c babel compiles escape sequences to the character anyway, so there isn't anything to be gained there unless the editor can't handle it i guess

**Comment by taion** [comment]
i'm ok with saying "let's use unicode in source"...

are you aware of any potential issues with using unicode instead of HTML entities in text for react, though?

File: index.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
+const { inspect } = require('util');
```
Comment (taion) [inline index.js:1]: wait if you're using `inspect` do you still need the custom formatting below?

Comment (jquense) [reply index.js:1]: just using inspect instead of `JSON.stringify` to help collapse large objects, etc

**Review by taion — APPROVED** [review]

**Pull request merged**